### ISSUE TITLE
[MNG-7578] Building Linux image on Windows impossible (patch incuded)

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8941]
**[Eugen Labun](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER297373)** opened **[MNG-7578](https://issues.apache.org/jira/browse/MNG-7578?redirect=false)** and commented

If you try to find `javac` in a Linux JDK using `Toolchain.findTool()` method, this will fail when the build is running on Windows, since the implementation [JavaToolchainImpl#findTool()](https://github.com/apache/maven/blob/maven-3.8.6/maven-core/src/main/java/org/apache/maven/toolchain/java/JavaToolchainImpl.java#L74-L86) appends ".exe" to the toolName (causing `javac.exe` is not found):

```java
 private static File findTool( String toolName, File installFolder )
 {
     File bin = new File( installFolder, "bin" ); //NOI18N
     if ( bin.exists() )
     {
         File tool = new File( bin, toolName + ( Os.isFamily( "windows" ) ? ".exe" : "" ) ); // NOI18N
         if ( tool.exists() )
         {
             return tool;
         }
     }
     return null;
}
```

The current workaround is to manually add a fake `javac.exe` file to the JDK `bin` directory. See [tool chain issue (building linux image on windows machine)](https://github.com/moditect/moditect/issues/107) in moditect project.

The `findTool` method could yet easily be changed to search for exact `toolName` as requested, with a fallback to `toolName.exe` for backward compatibility:

```java
 private static File findTool( String toolName, File installFolder )
 {
     File bin = new File( installFolder, "bin" );
     if ( bin.exists() )
     {
         File tool = new File( bin, toolName );
         if ( tool.exists() )
         {
             return tool;
         }
         File toolExe = new File( bin, toolName + ".exe" );
         if ( toolExe.exists() )
         {
             return toolExe;
         }
     }
     return null;
}
```

This would solve the problem.

---

**Affects:** 3.8.6

**Remote Links:**
- [GitHub Pull Request #861
](https://github.com/apache/maven/pull/861)

**Backported to:** [4.0.0-alpha-3](https://github.com/apache/maven/milestone/1?closed=1), [3.9.0](https://github.com/apache/maven/milestone/2?closed=1), [3.8.7](https://github.com/apache/maven/milestone/108?closed=1)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Makes sense, this also applies to other executalbes. I wonder why we need `.exe` at all to execute. It is implied on Windows anyway.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [a6b8a9e9fb3ca1f742883abd8cafe9eb6bf39691](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=a6b8a9e9fb3ca1f742883abd8cafe9eb6bf39691) for `maven-3.9.x` branch.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [a020540c5f5d5b3a01b08b1014465b3999f2be76](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=a020540c5f5d5b3a01b08b1014465b3999f2be76) for `maven-3.8.x` branch.
